### PR TITLE
Add COW support for MPS backend.

### PR DESCRIPTION
**Pull request by kulinseth**
Implemented COW logic w.r.t SharedMemoryStorage for MPSTensor.to(cpu), and CPUTensor.to(mps), so no explicit copy is needed unless a write occurred.

Removed the MANAGED UsageFlags and related logic. Removed PRIVATE_SMALL pool type and associated related logic to SHARED_SMALL.

### CODE REVIEW

File: aten/src/ATen/mps/MPSAllocator.mm [hunk 1]
```diff
@@ -511,12 +509,44 @@
...
+  }
+
+  std::lock_guard<std::recursive_mutex> lock(m_mutex);
+  BufferBlock* buffer_block = get_allocated_buffer_block(cpu_ptr);
+  if (buffer_block) {
+    // Already registered CPU pointer.
+    TORCH_INTERNAL_ASSERT(buffer_block->is_cpu_backed);
+    return buffer_block->buffer;
+  } else {
+
```
Comment (malfet) [inline aten/src/ATen/mps/MPSAllocator.mm:524]:
```suggestion
```

File: test/test_mps.py [hunk 2]
```diff
@@ -7383,7 +7383,7 @@ def helper(value, dim, index, idx_dtype=torch.int32):
     def test_embedding_dense_backward(self):
         def helper(n, d, m, idx):
             embeddingMPS = nn.Embedding(n, d, max_norm=True, device='mps')
-            emedding_weight = embeddingMPS.weight.detach().cpu()
+            embedding_weight = embeddingMPS.weight.detach().cpu()
```
Comment (malfet) [inline test/test_mps.py:7386]: It would be nice if this could be landed separately..

**Review by malfet — COMMENTED** [review]
Why lazy clone changes are needed?
Also, how can we tested it?

File: tools/autograd/derivatives.yaml [hunk 3]
```diff
@@ -451,9 +451,9 @@
   self: grad
   result: auto_linear
 
-- name: _lazy_clone(Tensor self) -> Tensor
-  self: grad
-  result: auto_linear
+- name: _lazy_clone(Tensor self, *, Device? device=None) -> Tensor
+  self: _lazy_clone_backward(grad, self.options())
+  result: _lazy_clone(self_t, device)
```
[the remark this answers is not in the document]
Comment (erpang007chenfs) [reply tools/autograd/derivatives.yaml:456]: I thought this is the right way to make sure the gradient flows to the target tensor. I am unfamiliar with this part of the code, do you have a suggestion on how I should change it?

File: aten/src/ATen/mps/MPSAllocator.mm (hunk 1 above)
Comment (erpang007chenfs) [reply aten/src/ATen/mps/MPSAllocator.mm:524]: Removed

File: test/test_mps.py (hunk 2 above)
Comment (erpang007chenfs) [reply test/test_mps.py:7386]: Reverted

File: c10/core/impl/COW.cpp [hunk 4]
```diff
@@ -129,9 +130,34 @@ C10_API void materialize_cow_storage(StorageImpl& storage) {
...
+        // Different devices copy
+        // TODO(Frank): Technically, it's possible to have CPUTensor owns
+        // MPS buffer, but it will be troublesome if we want to do other things.
+        // Eg. COW-again with the CPUTensor owning the MPS buffer, which means
+        // we will have logic to determine whether the data owned by the Tensor
+        // is created by the device it's on. It's possible, but will be
+        // much complicated, since the current assumption is, if the device
+        // is not COW'ed, data_ptr.get() == data.get().
+        // For now we will perform a copy.
+        new_data_ptr = storage.allocator()->clone(data_ptr.get(), storage.nbytes());
```
Comment (kurtamohler) [inline c10/core/impl/COW.cpp:153]: So in the case where `mps_tensor = cpu_tensor.to("mps")` is followed by a write to `cpu_tensor` and then a write to `mps_tensor`, both writes will initiate a copy, right? In other words, we're replacing the single immediate copy with two lazy copies. Do the benefits outweigh the costs?

Comment (erpang007chenfs) [reply c10/core/impl/COW.cpp:153]: This is correct case where the benefits will outweigh the cost. Do you have any suggestions on how this problem could be resolved other than let the `mps_tensor` owning the cpu memory? Or do you think we should figure out a way to support tensor owning a chunk of memory that's not created by the device allocator?

Comment (kurtamohler) [reply c10/core/impl/COW.cpp:153]: First, maybe it's ultimately ok to have the extra copy--I don't know. It depends on what the typical use case is expected to be.

> Do you have any suggestions on how this problem could be resolved other than let the mps_tensor owning the cpu memory?

I have a thought, though I'm not sure if it will be a good idea in the end. In the specific case I gave above, during the first write, to `cpu_tensor`, we could copy the CPU data to MPS and make `mps_tensor` point to it. Then `cpu_tensor` would take ownership of the CPU data when it materializes. Then, when `mps_tensor` is written, the data would already be on MPS, so it can materialize without copy.

A potential way to implement that would be to add a layer of indirection between `UniqueVoidPtr` and the underlying data, let's call it `Buffer`. After `mps_tensor = cpu.tensor("mps")` is called, `cpu_tensor` and `mps_tensor` would initially point to the same `Buffer`, and the buffer would point to the original CPU data. When `cpu_tensor` is written, it would copy the CPU data to MPS and change `Buffer` to point to the new MPS data.

But this is just a brainstorm, maybe there is a better way. This idea would also have to handle more complex cases, with more than just two tensors.

> Or do you think we should figure out a way to support tensor owning a chunk of memory that's not created by the device allocator?

It seems like that would be kind of against the whole purpose of the the `mps_tensor = cpu_tensor.to("mps")` call, wouldn't it? To me it seems like it would be odd for `mps_tensor` to still point to CPU data even after materialization. But I don't know, is it really possible/desirable to do that?

**Comment by zwshan** [comment]
Hello, I'm very excited about PyTorch potentially supporting unified memory on macOS. May I ask approximately how much longer it will take to be merged into the main branch?

**Comment by skotapati** [comment]
Time to take another crack at merging this, it's been a little bit so I'm re-requesting review. I'd also like to rebase this on top of https://github.com/pytorch/pytorch/pull/142202 for additional test coverage, so any help merging that PR would be appreciated.

**Comment by creatica-soft** [comment]
Adding unified memory support would help increase performance of training AI models on M1/M2 devices. As of now, the data transfer from CPU to MPS takes about 80% of the training loop and only 20% is spent on GPU training. The overhead of copying memory from CPU to MPS is too big! Another approach would be disclosing/documenting how to write to MPS allocated tensor's memory directly from CPU threads to avoid the bottleneck with memory copying. As of now, trying to do this leads to heap-memory buffer overflows.

**Comment by michal1000w** [comment]
Are any updates on that? I agree with @creatica-soft that unified memory for PyTorch on maps would be a huge deal and could potentially improve the performance by a lot.

**Comment by kurtamohler** [comment]
I've been working on a stack of PRs #148408 which will add unified memory support for MPS
